test/mpi/main.cc

```diff
@@ -383,34 +397,103 @@ void ProcessArgs(int argc, char** argv)
     }
 }
 
+#ifdef HAVE_CUDA
+void cuda_set_device()
+{
+    MPI_Comm local_comm;
+    int cuda_dev_count;
+    int local_rank;
+    MPI_Comm_split_type(MPI_COMM_WORLD, OMPI_COMM_TYPE_NODE, 0, MPI_INFO_NULL,
+                        &local_comm);
+    MPI_Comm_rank(local_comm, &local_rank);
+    CUDA_CHECK(cudaGetDeviceCount(&cuda_dev_count));
+    if(local_rank >= cuda_dev_count) {
+        std::cerr << "*** UCC TEST FAIL: "
+                  << "not enough CUDA devices on the node to map processes.\n";
+        MPI_Abort(MPI_COMM_WORLD, -1);
+    }
+    CUDA_CHECK(cudaSetDevice(local_rank));
+}
+#endif
+
 int main(int argc, char *argv[])
 {
     std::chrono::steady_clock::time_point begin =
         std::chrono::steady_clock::now();
     int rank;
     int failed = 0;
-    UccTestMpi test(argc, argv, UCC_THREAD_SINGLE);
+    ucc_thread_mode_t tm = UCC_THREAD_SINGLE;
+    int required = (tm == UCC_THREAD_SINGLE) ? MPI_THREAD_SINGLE
+                                             : MPI_THREAD_MULTIPLE;
+    int size, provided;
+    ucc_lib_h lib;
+    ucc_context_h ctx;
+    ucc_lib_config_h lib_config;
+    ucc_context_config_h ctx_config;
+    UccTestMpi *test;
+
+    /* Init ucc library */
+    ucc_lib_params_t lib_params = {
+        .mask = UCC_LIB_PARAM_FIELD_THREAD_MODE,
+        .thread_mode = tm,
+        /* .coll_types = coll_types, */
+    };
+
+    /* Init ucc context for a specified UCC_TEST_TLS */
+    ucc_context_params_t ctx_params = {
+        .mask     = UCC_CONTEXT_PARAM_FIELD_TYPE,
+        .ctx_type = UCC_CONTEXT_EXCLUSIVE,
+    };
+    UCC_CHECK(ucc_lib_config_read(NULL, NULL, &lib_config));
+    UCC_CHECK(ucc_init(&lib_params, lib_config, &lib));
```

Comment:
Need to rebase carefully: CONTEXT_EXCLUSIVE is not used any more (for multithreaded mode); 
Why you extracted all these UCC setup out of test constructor?
In master ParseArgs currently is made BEFORE ucc_test, so that all parameters are available for constructor.

Reply:
fixed